4단계 - 로또(수동)

julee-0430 opened the pull request:
안녕하세요!

이번 미션에서는 아래 내용을 고민해본 것 같습니다. 
- 수동 로또와 자동 로또 개수를 관리할 LottoPurchasePlan 객체 추가
- 원시값 포장을 위한 Count 객체 추가

각 객체의 역할에 이런 메서드가 추가될 수 있나? 를 계속 고민했던 것 같은데 잘 되었는지는 모르겠네요..ㅎㅎ
감사합니다!

julee-0430 committed:
[refactor] LottoNumber 객체 재사용으로 메모리 사용량 고려

- LottoNumber 객체는 범위가 지정되어 있으므로 매번 새로 생성하지 않고 인스턴스를 재사용하도록 변경

julee-0430 committed:
[refactor] LottoResultCalculator 에서 최종 결과 계산

- Lottos 에 있던 계산 로직을 LottoResultCalculator로 이동

julee-0430 committed:
[refactor] LottoNumber 내부 정적 팩토리 메서드로 변경

julee-0430 committed:
[refactor] LottoResultCalculator 대신 Lottos 내부에서 반복하도록 변경

julee-0430 committed:
[refactor] Lotto Count 객체 (원시값 포장) 추가

julee-0430 committed:
[refactor] 누락 테스트 (LottosTest) 반영 및 readme 업데이트

julee-0430 committed:
[feat] 수동 로또와 자동 로또 개수를 관리할 LottoPurchasePlan 객체 추가

julee-0430 committed:
[refactor] LottoApplication 에서 LottoController 로 분리

julee-0430 commented on src/test/java/lotto/model/LottosTest.java:
이 파일은 지난 번 피드백 때 LottoResultCalculator 클래스가 굳이 필요없을 것 같다는 피드백을 반영하는 과정에서 누락되었네요..ㅎㅎ

javajigi commented on src/main/java/lotto/model/LottoPurchasePlan.java:
```diff
@@ -0,0 +1,46 @@
+package lotto.model;
+
+public class LottoPurchasePlan {
```
3단계에서 4단계로 요구사항이 변경될 때 로또를 생성하는 부분의 요구사항만 변경됐다.
로또를 생성하는 부분과 로또 결과를 구하는 부분을 다음과 같은 구조의 인터페이스로 분리해 보는 연습을 해보면 어떨까?
이와 같이 인터페이스로 구현했을 때의 잇점에 대해 고민해 보는 시간을 가져본다.

```java
public interface LottosGenerator {
    Lottos generate();
}
```

javajigi reviewed (changes requested):
3단계 구조의 변화를 최소화하면서 4단계 요구사항 반영했네요. 👍 
로또 생성과 관련해 인터페이스 기반으로 구현하는 연습을 해봤으면 하는 바람으로 피드백 하나 남겨 봅니다.
이 피드백을 포함해 다른 부분에 리팩터링할 부분은 없는지 찾아보면 좋겠습니다.

julee-0430 committed:
[refactor] LottosGenerator 로 로또 생성기 로직 추상화 및 구현체 추가

julee-0430 committed:
[refactor] Lotto 생성 로직 LottoFactory로 위임

- ManualLottosGenerator는 수동 로또 티켓들(Lottos)을 발행하는 클래스라서 Lotto 생성은 LottoFactory 로 이동
- split 이나 trim 등의 문자열 처리 로직이 포함되어 있어 Lotto 생성자 대신 팩토리 패턴 사용

julee-0430 committed:
[refactor] View 에서 도메인 인스턴스를 바로 반환하지 않고 Controller 에서 생성하도록 둠.

- 원시값 포장 객체는 View 에서 생성하도록 유지

julee-0430 committed:
[refactor] 누락 테스트 반영

julee-0430 committed:
[refactor] LottoFactory 대신 Lotto 내부 정적 팩터리 메서드 사용

julee-0430 committed:
[refactor] InputView 에서 원시값만 반환하도록

julee-0430 commented:
확실히 인터페이스로 구현하니 자동 방식과 수동 방식이 뚜렷하게 보이고, 언제든지 바꿔서 구현할 수 있는 부분이 장점 같습니다!

다만 이렇게 하니 구현체가 -er 네이밍을 갖게 되면서 LottoPurchasePlan의 autoCount 값을 직접 꺼내는 형태로 구현된 것 같아 좀 신경이 쓰입니다..


추가적으로는 InputView에서 원시값이 아닌 도메인 인스턴스들을 생성해서 Controller 로 반환하는 게 맘에 안 들었는데 그 부분도 수정해보았습니다!

javajigi commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,37 @@
+        List<String> manualInputs = InputView.readMaualLottoInput(lottoPurchasePlan.manualCount().value());
+        Lottos manuals = new ManualLottosGenerator(manualInputs).generate();
+        Lottos autos = new AutoLottosGenerator(lottoPurchasePlan).generate();
+        return Lottos.of(manuals, autos);
```
이와 같이 인터페이스를 구현하고, 구현체가 서로 연결되어 있는 경우 아래와 같은 구현체를 통해 구현체를 통합할 수 있음.
보통 디자인 패턴에서 컴포지트 패턴으로 알려져 있음.
아래와 같은 객체 추가에 따른 효과를 느껴 봤으면 하는 바람으로 피드백 남겨봄.

```java
public class LottosBundleGenerator implements LottosGenerator {
    private final Money money;
    private final List<String> manualLottoText;

    public LottosBundleGenerator(Money money, List<String> manualLottoText) {
        this.money = money;
        this.manualLottoText = manualLottoText;
    }

    @Override
    public Lottos generate() {
        // ManualLottosGenerator 활용해 수동 로또 생성
        // AutoLottosGenerator 활용해 자동 로또 생성(수동 로또 수 만큼 Money 차감)
        return 두 로또를 합쳐서 반환;
    }
}
```

javajigi commented on src/main/java/lotto/model/LottoPurchasePlan.java:
```diff
@@ -0,0 +1,42 @@
+package lotto.model;
+
+public class LottoPurchasePlan {
```
인터페이스 기반으로 구현하면서 이 객체가 반드시 필요한 객체인지 한번 고민해 보는 시간을 가지면 어떨까?
AI를 통해 피드백을 받아봐도 좋겠다.

javajigi reviewed (changes requested):
인터페이스 기반으로 개발한 효과를 느껴봤기를 바랍니다. 👍 
이와 같은 인터페이스 구현체가 서로 연결되어 있을 경우 추가로 고려해 봤으면 하는 피드백이 있어 남겨봄

julee-0430 committed:
[refactor] LottosBundleGenerator로 컴포지트 패턴 사용 후 불필요한 클래스 제거

julee-0430 committed:
[refactor] 불필요 import 제거

javajigi commented on src/main/java/lotto/model/LottosBundleGenerator.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public Lottos generate() {
+        Count manualCounts = new Count(manualLottoText.size());
+        Count autoCounts = money.countBuyableLottoTickets().subtract(manualCounts);
```
```suggestion
        Count autoCounts = money.countBuyableLottoTickets().subtract(manualLottoText.size());
```
subtract에 Count를 인자로 받는 메서드도 제공하지만, int를 인자로 받는 메서드를 오버로딩한다면 위와 같이도 구현 가능함.
생성자만 오버로딩하지 말고 메서드에 대한 오버로딩도 추가해 보는 습관을 들이면 좋겠음.

javajigi reviewed (approved):
피드백 반영 👍 
지금까지 로또 미션 진행하느라 수고했어요.
소소한 피드백 하나 남겼으니 생각해보는 시간 가지면 좋겠습니다.
